platform/linux-generic/pktio/mdev.c

```diff
@@ -0,0 +1,589 @@
+/* Copyright (c) 2018, Linaro Limited
+ * All rights reserved.
+ *
+ * SPDX-License-Identifier:     BSD-3-Clause
+ */
+
+#include "config.h"
+
+#ifdef ODP_MDEV
+
+#include <dirent.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include <linux/vfio.h>
+
+#include <odp_posix_extensions.h>
+#include <odp_align_internal.h>
+#include <odp_debug_internal.h>
+
+#include <odp/api/hints.h>
+
+#include <pktio/mdev.h>
+#include <pktio/uapi_net_mdev.h>
+
+/**
+ * returns a valid VFIO container
+ * fd must be closed by caller
+ */
+static int get_container(void)
+{
+	int container;
+	int ret;
+
+	/* Create a new container */
+	container = open("/dev/vfio/vfio", O_RDWR);
+	if (container < 0) {
+		ODP_ERR("Failed to create new VFIO container\n");
+		goto out;
+	}
+
+	ret = ioctl(container, VFIO_GET_API_VERSION);
+	if (ret != VFIO_API_VERSION) {
+		ODP_ERR("VFIO API version mismatch: expected %i, got %i\n",
+			VFIO_API_VERSION, ret);
+		goto out;
+	}
+
+	ret = ioctl(container, VFIO_CHECK_EXTENSION, VFIO_TYPE1_IOMMU);
+	if (!ret) {
+		ODP_ERR("Container doesn't support VFIO_TYPE1_IOMMU\n");
+		goto out;
+	}
+
+	return container;
+
+out:
+	if (container != -1)
+		close(container);
+	return -1;
+}
+
+/**
+ * returns a valid VFIO group
+ * fd must be closed by caller
+ */
+static int get_group(int grp_id)
+{
+	char path[32];
+	int group;
+	int ret;
+	struct vfio_group_status group_status = {
+		.argsz = sizeof(group_status)
+	};
+
+	snprintf(path, sizeof(path), "/dev/vfio/%d", grp_id);
+	group = open(path, O_RDWR);
+	if (group < 0) {
+		ODP_ERR("Failed to open %s: %s\n", path, strerror(errno));
+		goto out;
+	}
+
+	ret = ioctl(group, VFIO_GROUP_GET_STATUS, &group_status);
+	if (ret < 0) {
+		ODP_ERR("Failed to get group status\n");
+		goto out;
+	}
+
+	/* Test the group is viable and available */
+	if (!(group_status.flags & VFIO_GROUP_FLAGS_VIABLE)) {
+		ODP_ERR("Group is not viable\n");
+		goto out;
+	}
+
+	return group;
+
+out:
+	if (group != -1)
+		close(group);
+	return -1;
+}
+
+static void vfio_find_sparse_mmaps(struct vfio_info_cap_header *hdr,
+				   struct vfio_region_info_cap_sparse_mmap
+				   **sparse)
+{
+	*sparse =
+	    odp_container_of(hdr, struct vfio_region_info_cap_sparse_mmap,
+			     header);
+}
+
+static struct vfio_info_cap_header *
+vfio_get_region_info_cap(struct vfio_region_info *info, __u16 id)
+{
+	struct vfio_info_cap_header *hdr;
+	void *ptr = info;
+
+	if (!(info->flags & VFIO_REGION_INFO_FLAG_CAPS))
+		return NULL;
+
+	for (hdr =
+	     (struct vfio_info_cap_header *)((char *)ptr + info->cap_offset);
+	     hdr != ptr;
+	     hdr = (struct vfio_info_cap_header *)((char *)ptr + hdr->next)) {
```

Comment:
Specifying `./configure --enable-mdev-support=yes CC=clang` results in lines 130 and 132 being flagged by clang:
```
  CC       pktio/mdev.lo
pktio/mdev.c:130:7: error: cast from 'char *' to 'struct vfio_info_cap_header *'
      increases required alignment from 1 to 4 [-Werror,-Wcast-align]
             (struct vfio_info_cap_header *)((char *)ptr + info->cap_offset);
             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pktio/mdev.c:132:13: error: cast from 'char *' to
      'struct vfio_info_cap_header *' increases required alignment from 1 to 4
      [-Werror,-Wcast-align]
             hdr = (struct vfio_info_cap_header *)((char *)ptr + hdr->next)) {
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 errors generated.
Makefile:1222: recipe for target 'pktio/mdev.lo' failed
make[1]: *** [pktio/mdev.lo] Error 1
```

Reply:
I've fixed clang compilation.

I guess if we insert compilation test into .travis.yml without proper Ubuntu 17.10 docker images -- it will fail on every upcoming PR which doesn't make much sense. I believe it shall be in separate PR accompanied with all prerequisite changes in Travis images. What do you think ?